spot_driver_cpp/src/api/default_spot_api.cpp

```diff
@@ -0,0 +1,74 @@
+#include <spot_driver_cpp/api/default_time_sync_api.hpp>
+
+namespace spot_ros2 {
+
+DefaultSpotApi::DefaultSpotApi(const std::string& sdk_client_name)
+    : client_sdk_{::bosdyn::client::CreateStandardSDK(sdk_client_name)} {}
+
+tl::expected<void, std::string> DefaultSpotApi::createRobot(const std::string& ip_address,
+                                                            const std::string& robot_name) {
+  robot_name_ = robot_name;
+
+  auto create_robot_result = client_sdk_->CreateRobot(ip_address);
+  if (!create_robot_result.status) {
+    return tl::make_unexpected("Received error result when creating SDK robot interface: " +
+                               create_robot_result.status.DebugString());
+  }
+
+  robot_ = std::move(create_robot_result.response);
+
+  return {};
+}
+
+tl::expected<void, std::string> DefaultSpotApi::authenticate(const std::string& username, const std::string& password) {
+  const auto authenticate_result = robot_->Authenticate(username, password);
+  if (!authenticate_result) {
+    return tl::make_unexpected("Authentication with provided username and password did not succeed.");
+  }
+  // Start time synchronization between the robot and the client system.
+  // This must be done only after a successful authentication.
+  const auto start_time_sync_response = robot_->StartTimeSync();
+  if (!start_time_sync_response) {
+    return tl::make_unexpected("Failed to start time synchronization.");
+  }
+
+  const auto get_time_sync_thread_response = robot_->GetTimeSyncThread();
+  if (!get_time_sync_thread_response) {
+    return tl::make_unexpected("Failed to get the time synchronization thread.");
+  }
+  time_sync_api_ = std::make_shared<DefaultTimeSyncApi>(get_time_sync_thread_response.response);
+
+  return {};
+}
+
+tl::expected<bool, std::string> DefaultSpotApi::hasArm() const {
+  // Determine if Spot has an arm by checking if the client for gripper camera parameters exists, since Spots without
+  // arms do not have this client.
+  const auto list_result = robot_->ListServices();
+  if (!list_result.status) {
+    return tl::make_unexpected("Failed to retrieve list of Spot services.");
+  }
+
+  const auto& services = list_result.response;
+
+  return std::find_if(services.cbegin(), services.cend(), [](const ::bosdyn::api::ServiceEntry& entry) {
+           return entry.name() == ::bosdyn::client::GripperCameraParamClient::GetDefaultServiceName();
+         }) != services.cend();
+}
+
+tl::expected<std::unique_ptr<ImageClientApi>, std::string> DefaultSpotApi::imageClient() const {
+  const auto image_client_result = robot_->EnsureServiceClient<::bosdyn::client::ImageClient>(
+      ::bosdyn::client::ImageClient::GetDefaultServiceName());
+  if (!image_client_result.status) {
+    return tl::make_unexpected("Failed to create image client.");
+  }
+
+  return std::make_unique<DefaultImageClientApi>(image_client_result.response, time_sync_api_, robot_name_);
```

Comment:
This has to do with a previous comment. Currently, the `::bosdyn::client::ImageClient` is held by the robot, and it is shared to the external world using a raw pointer.

If we call this method twice, and we take ownership of the created object using a `unique_ptr` (which we do), this creates a potential memory leak. Destroying any of the two `unique_ptr`, will destroy the object. 

The `imageClient` method is not a factory, because part of the object is held by the robot.

I suggest the following:

- add a private `shared_ptr<ClientApi>` field to the class `DefaultSpotApi`.
- update the `image_client` method to return a  `shared_ptr<ClientApi>`.
- update the `image_client` method to create a `shared_ptr<ClientApi>` only if it has not been created before.
- with this change, the ClientApi is destroyed only once, so it is safe to convert the raw pointer into a unique_ptr<::bosdyn::client::ImageClient> and pass it to the constructor of the `DefaultClientApi`. This will also be destroyed once.

Reply:
```cpp
tl::expected<std::shared_ptr<ImageClientApi>, std::string> DefaultSpotApi::imageClient() const {
  if (!image_client_) {
    const auto image_client_result = robot_->EnsureServiceClient<::bosdyn::client::ImageClient>(
        ::bosdyn::client::ImageClient::GetDefaultServiceName());
    if (!image_client_result.status) {
      return tl::make_unexpected("Failed to create image client.");
    }
    auto bdai_client = std::make_unique<::bosdyn::client::ImageClient>(image_client_result.response);
    image_client_ = std::make_shared<DefaultImageClientApi>(std::move(bdai_client), time_sync_api_, robot_name_);
  } else {
    return image_client_;
  }
}
```